calyx/src/passes/infer_share.rs

```diff
@@ -0,0 +1,386 @@
+
+        let state_shareable = ShareSet::from_context::<true>(ctx);
+
+        let mem_ids = HashSet::from([
+            ir::Id::new("std_mem_d1", None),
+            ir::Id::new("std_mem_d2", None),
+            ir::Id::new("std_mem_d3", None),
+            ir::Id::new("std_mem_d4", None),
+        ]);
+
+        Ok(InferShare {
+            print_dmap: opts[0],
+            state_shareable,
+            no_share: mem_ids,
+            main: ctx.entrypoint.clone(),
+        })
+    }
+
+    fn clear_data(&mut self) {}
+}
+
+impl Visitor for InferShare {
+    fn require_postorder() -> bool {
+        true
+    }
+    fn start(
+        &mut self,
+        comp: &mut ir::Component,
+        _sigs: &ir::LibrarySignatures,
+        _comps: &[ir::Component],
+    ) -> VisResult {
+        //if the component is main, then we can stop checking
+        if comp.name == self.main {
+            return Ok(Action::Continue);
+        }
+
+        //cannot contain any ref cells, non-shareable components, or mem cells
+        if comp.cells.iter().any(|cell| {
+            let cell_ref = cell.borrow();
+            cell_ref.is_reference()
+                || match cell_ref.type_name() {
+                    Some(name) => self.no_share.contains(name),
+                    None => false,
+                }
+        }) {
+            self.no_share.insert(comp.name.clone());
+            return Ok(Action::Continue);
+        }
+
+        //build the domination map
+        let mut dmap = DominatorMap::new(
+            &mut comp.control.borrow_mut(),
+            comp.name.id.clone(),
+        );
+        if self.print_dmap {
+            println!("{dmap:?}");
+        }
+
+        for (node, dominators) in dmap.map.iter_mut() {
+            //get the reads
+            let mut reads: ReadSet = ReadSet::default();
+            if let Some(c) =
+                DominatorMap::get_control(*node, &comp.control.borrow())
+            {
+                reads.get_reads_from_control(
+                    c,
+                    &self.state_shareable,
+                    is_begin_id(c, *node),
+                );
+            }
+
+            //if read and write occur in same group/invoke, then we cannot label it
+            //shareable. So we remove node from its dominators
+            dominators.remove(node);
+            for cell_name in reads.reads.clone() {
+                let key = NameSearch::new(cell_name);
+                if !key.is_written(dominators, comp, &self.state_shareable) {
+                    self.no_share.insert(comp.name.clone());
+                    return Ok(Action::Continue);
+                }
+            }
+        }
+        comp.attributes.insert("state_share", 1);
+        self.state_shareable.add(comp.name.clone());
+        Ok(Action::Continue)
+    }
+}
+
+///Contains the ids of all the cells that are read from in a given "node" in
+///the domination map graph
+#[derive(Default)]
+struct ReadSet {
```

Comment:
If this struct is used to manage the DominatorMap then it should be not be in the `infer_share` file